sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -452,6 +452,21 @@ class Analyzer(
 
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
       case i @ InsertIntoTable(u: UnresolvedRelation, parts, child, _, _) if child.resolved =>
+        // A partitioned relation's schema can be different from the input logicalPlan, since
+        // partition columns are all moved after data columns. We Project to adjust the ordering.
+        val inputWhenPartsNonEmpty = if (parts.nonEmpty) {
+          val (inputPartCols, inputDataCols) = child.output.partition { attr =>
+            parts.contains(attr.name)
+          }
+          if (child.output == (inputDataCols ++ inputPartCols)) {
+            child
+          } else {
+            Project(inputDataCols ++ inputPartCols, child)
```

Comment:
Yea, I see. I did this before but two tests in `HiveQuerySuite` will be failed due to this because it tests against the analyzed logical plan...

If we always do project here, we may need to modify `HiveQuerySuite`. Let me do this and then you see if it works for you. Thanks!